File: golang/finite_fields.go
package base

import (
	"fmt"
	"math"
)

type FieldElementStruct struct {
	Num   int
	Prime int
}

// utils
func extendedEuclideanAlgorithm(a, b int) (int, int, int) {
	if b == 0 {
		return 1, 0, a
	}
	x, y, gcd := extendedEuclideanAlgorithm(b, a%b)
	return y, x - (a/b)*y, gcd
}

func Inverse(a int, p int) int {
	x, _, gcd := extendedEuclideanAlgorithm(int(a), p)
	if gcd != 1 {
		// a has no modular inverse
		return 0
	}
	return (x%p + p) % p
}

// finite field methods
func FieldElement(num int, prime int) (*FieldElementStruct, error) {
	if num > prime || num < 0 {
		return nil, fmt.Errorf("num %d not in field range 0 to %d", num, prime-1)
	}
	return &FieldElementStruct{Num: num, Prime: prime}, nil
}

func (f *FieldElementStruct) Equals(other *FieldElementStruct) bool {
	return f.Num == other.Num && f.Prime == other.Prime
}

func (f *FieldElementStruct) NotEquals(other *FieldElementStruct) bool {
	return f.Num != other.Num || f.Prime != other.Prime
}

func (f *FieldElementStruct) Add(other *FieldElementStruct) (*FieldElementStruct, error) {
	if f.Prime != other.Prime {
		return nil, fmt.Errorf("cannot add two numbers in different Fields")
	}
	sum := f.Num + other.Num
	for sum < 0 {
		sum += f.Prime
	}
	return &FieldElementStruct{Num: sum % f.Prime, Prime: f.Prime}, nil
}

func (f *FieldElementStruct) Sub(other *FieldElementStruct) (*FieldElementStruct, error) {
	if f.Prime != other.Prime {
		return nil, fmt.Errorf("cannot sub two numbers in different Fields")
	}
	sum := f.Num - other.Num
	for sum < 0 {
		sum += f.Prime
	}
	return &FieldElementStruct{Num: sum % f.Prime, Prime: f.Prime}, nil
}

func (f *FieldElementStruct) Mul(other *FieldElementStruct) (*FieldElementStruct, error) {
	if f.Prime != other.Prime {
		return nil, fmt.Errorf("cannot multiple two numbers in different Fields")
	}
	sum := f.Num * other.Num
	for sum < 0 {
		sum += f.Prime
	}
	return &FieldElementStruct{Num: sum % f.Prime, Prime: f.Prime}, nil
}

func (f *FieldElementStruct) Pow(index int) (*FieldElementStruct, error) {
	result := int(math.Pow(float64(f.Num), float64(index)))
	for result < 0 {
		result += f.Prime
	}
	return &FieldElementStruct{Num: result % f.Prime, Prime: f.Prime}, nil
}

func (f *FieldElementStruct) Truediv(other *FieldElementStruct) (*FieldElementStruct, error) {
	if f.Prime != other.Prime {
		return nil, fmt.Errorf("cannot multiple two numbers in different Fields")
	}
	index := f.Prime - 2
	result := f.Num * int(math.Pow(float64(other.Num), float64(index)))
	return &FieldElementStruct{Num: result, Prime: f.Prime}, nil
}

File: python/finite_fields.py
class FieldElement:
  def __init__(self, num, prime):
    if num >= prime or num < 0:
      error = 'Num {} not in field range 0 to {}'.format(num, prime - 1)
      raise ValueError(error)
    
    self.num = num
    self.prime = prime
  
  # Output meaningful string when print class object
  def __repr__(self):
    return 'FieldElement_{}({})'.format(self.prime, self.num)
  
  def __eq__(self, other):
    if other is None:
      return False
    return self.num == other.num and self.prime == other.prime
  
  def __ne__(self, other):
    if self.num != other.num or self.prime != other.prime:
      return True
    return False
  
  def __add__(self, other):
    if self.prime != other.prime:
        raise TypeError('Cannot add two numbers in different Fields')
    num = (self.num + other.num) % self.prime 
    return self.__class__(num, self.prime)

  def __sub__(self, other):
    if self.prime != other.prime:
        raise TypeError('Cannot add two numbers in different Fields')
    num = (self.num - other.num) % self.prime 
    return self.__class__(num, self.prime)
  
  def __mul__(self, other):
    if self.prime != other.prime:
        raise TypeError('Cannot multiply two numbers in different Fields')
    num = (self.num * other.num) % self.prime
    return self.__class__(num, self.prime)
  
  def __rmul__(self, coefficient):
    num = (self.num * coefficient) % self.prime
    return self.__class__(num=num, prime=self.prime)

  def __pow__(self, exponent):
    n = exponent % (self.prime - 1)
    num = pow(self.num, n, self.prime)
    return self.__class__(num, self.prime)

  def __truediv__(self, other):
    if self.prime != other.prime:
      raise TypeError('Cannot divide two numbers in different Fields')

    num = (self.num * pow(other.num, self.prime - 2, self.prime)) % self.prime
    return self.__class__(num, self.prime)

a = FieldElement(2, 31)
b = FieldElement(2, 31)
c = FieldElement(15, 31)
